components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/SimpleObjectEncoder.scala

```diff
@@ -1,47 +1,101 @@
 package pl.touk.nussknacker.engine.api.typed
 
-import cats.data.{Validated, ValidatedNel}
-import cats.implicits.catsSyntaxValidatedId
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+import cats.implicits.{catsSyntaxValidatedId, toTraverseOps}
+import io.circe.Json.{fromBigDecimal, fromBigInt, fromBoolean, fromDouble, fromFloat, fromInt, fromLong, fromString}
 import io.circe.{ACursor, Decoder, DecodingFailure, Json}
-import io.circe.Json.{fromBoolean, fromDouble, fromFloat, fromInt, fromLong, fromString}
-import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedClass}
+import pl.touk.nussknacker.engine.api.typed.typing._
+
+import java.math.BigInteger
+import scala.jdk.CollectionConverters._
 
 // TODO: Add support for more types.
 object SimpleObjectEncoder {
-  private val intClass     = Typed.typedClass[Int]
-  private val longClass    = Typed.typedClass[Long]
-  private val floatClass   = Typed.typedClass[Float]
-  private val doubleClass  = Typed.typedClass[Double]
-  private val booleanClass = Typed.typedClass[Boolean]
-  private val stringClass  = Typed.typedClass[String]
-
-  def encode(typ: TypedClass, data: Any): ValidatedNel[String, Json] = (typ, data) match {
-    case (`intClass`, intValue: Int) =>
-      fromInt(intValue).validNel
-    case (`longClass`, longValue: Long) =>
-      fromLong(longValue).validNel
-    case (`floatClass`, floatValue: Float) =>
-      fromFloat(floatValue).map(_.validNel).getOrElse(s"Could not encode $floatValue as json.".invalidNel)
-    case (`doubleClass`, doubleValue: Double) =>
-      fromDouble(doubleValue).map(_.validNel).getOrElse(s"Could not encode $doubleValue as json.".invalidNel)
-    case (`booleanClass`, booleanValue: Boolean) =>
-      fromBoolean(booleanValue).validNel
-    case (`stringClass`, stringValue: String) =>
-      fromString(stringValue).validNel
-    case (klass, value) if value.getClass == klass.klass =>
-      s"No encoding logic for $typ.".invalidNel
-    case (klass, value) =>
-      s"Mismatched class and value: $klass and $value".invalidNel
+  private val intClass        = Typed.typedClass[Int]
+  private val shortClass      = Typed.typedClass[Short]
+  private val longClass       = Typed.typedClass[Long]
+  private val floatClass      = Typed.typedClass[Float]
+  private val doubleClass     = Typed.typedClass[Double]
+  private val booleanClass    = Typed.typedClass[Boolean]
+  private val stringClass     = Typed.typedClass[String]
+  private val byteClass       = Typed.typedClass[Byte]
+  private val bigIntegerClass = Typed.typedClass[BigInteger]
+  private val bigDecimalClass = Typed.typedClass[java.math.BigDecimal]
+
+  def encodeValue(value: Any): ValidatedNel[String, Json] = value match {
```

Comment:
Why it doesn't take the type anymore? It was redundant?

Reply:
yep, and it made it difficult to call recursively

Reply:
though maybe it wasn't 100% reduntant, now the case with mismatched type-value isn't handled the same :thinking: